pandas/core/internals.py

```diff
@@ -2451,6 +2462,182 @@ def shift(self, periods, axis=0, mgr=None):
+        raise on an except if raise == True
+        """
+        # if we are passed a datetime64[ns, tz]
+        if is_period_dtype(dtype):
+            dtype = PeriodDtype(dtype)
+            values = self.values
+            # ToDo: asfreq??
+            return self.make_block(values)
+
+        # delegate
+        return super(PeriodBlock, self)._astype(dtype=dtype, **kwargs)
+
+    def external_values(self):
+        """ we internally represent the data as a PeriodIndex, but for external
+        compat with ndarray, export as a ndarray of Period """
+        return self.values.asobject.values
+
+    def get_values(self, dtype=None):
+        # return object dtype as Period
+        if dtype == object:
+            infered = lib.map_infer(self.values.ravel(),
+                                    self.values._box_func)
+            return infered.reshape(self.values.shape)
+        return self.values
+
+    def to_dense(self):
+        # ToDo: FIXME
+        # called from Series.get_values()
+        return self.values.asobject
+
+    def _try_fill(self, value):
+        """ if we are a NaT, return the actual fill value """
+        if value is tslib.NaT or np.array(isnull(value)).all():
+            value = tslib.iNaT
+        elif isinstance(value, Period):
+            # Period Nat can be handled here
+            value = value.ordinal
+        elif is_integer(value):
+            # regarded as ordinal
+            pass
+
+        return value
+
+    def _try_coerce_args(self, values, other):
+        """
+        Coerce values and other to dtype 'i8'. NaN and NaT convert to
+        the smallest i8, and will correctly round-trip to NaT if converted
+        back in _try_coerce_result. values is always ndarray-like, other
+        may not be
+
+        Parameters
+        ----------
+        values : ndarray-like
+        other : ndarray-like or scalar
+
+        Returns
+        -------
+        base-type values, values mask, base-type other, other mask
+        """
+        values_mask = isnull(values)
+        values = values.view('i8')
+        other_mask = False
+
+        if isinstance(other, bool):
+            raise TypeError
+        elif is_null_datelike_scalar(other):
+            other = tslib.iNaT
+            other_mask = True
+        elif isinstance(other, (Period)):
+            # ToDo: compare freq
+            other_mask = isnull(other)
+            other = other.ordinal
+        elif hasattr(other, 'dtype') and is_integer_dtype(other):
+            other = other.view('i8')
+        else:
+            try:
+                other = np.asarray(other)
+                other_mask = isnull(other)
+
+                other = other.astype('i8', copy=False).view('i8')
+            except ValueError:
+
+                # coercion issues
+                # let higher levels handle
+                raise TypeError
+
+        return values, values_mask, other, other_mask
+
+    def _try_coerce_result(self, result):
```

Comment:
if you factor as I suggest above, you can prob make a
```
def _shallow_copy_index(result):
    return PeriodIndex(result, freq=self.values.freq)
```

to do the new internal index creation (and same for ``DatetimeTZBlock``) to make this logic simpler (maybe)